### PR TITLE
[RuntimeEnv, Windows] Fix working_dir, pip & conda for windows

### PR DESCRIPTION
**Pull request by jbedorf**
## Why are these changes needed?

Various options of the `runtime_env` method, when using Windows, are currently broken due to the changes in [this PR](https://github.com/ray-project/ray/pull/25659/files). That PR removed the use of the `command_prefix` in the context. This work restores the usage of that. However, due to the different launch methods between Linux and Windows the changes had to be made in multiple locations to ensure that lists are returned instead of strings. 

To prevent this breaking in the future various tests have been fixed/enabled. However, there are some lingering issues with the tests:
* The top level folder of the `working_dir` is not deleted due to not being able to delete a folder that is in use on Windows. This is accounted for in the tests by whitelisting that particular folder in various tests.
* For the `conda` environment option the deleting is not complete. Similar the deleting fails because a number of files and folders are in use. Typically they are in use by the `ray.util.client.server` process. As such not all tests are enabled for Windows as they would keep failing, and leave behind sizeable temporary files.

Other:
* Fixed issue where Ray and the temp folder are placed on different drives and as such the change directory method did not work.
* Fixed a number of tests that had recursion errors on Windows

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jbedorf** [comment]
@mattip Given that you previously worked on Windows related PRs could you help find some reviewers? Thanks!

**Comment by jbedorf** [comment]
@architkulkarni Can you help move this forward? Thanks!

File: python/ray/_private/runtime_env/pip.py
Comment (architkulkarni) [inline python/ray/_private/runtime_env/pip.py:507]: Should we make `get_virtualenv_activate_command` always return `List[str]` to avoid this if-else?

File: python/ray/_private/test_utils.py
Comment (architkulkarni) [inline python/ray/_private/test_utils.py:1251]: Based on the term "whitelist" I would expect subset instead of == here, what do you think?

**Review by architkulkarni — APPROVED** [review]
Thanks for the contribution!

Regarding the conda tests:
> As such not all tests are enabled for Windows as they would keep failing, and leave behind sizeable temporary files.

Certainly if they're failing we shouldn't enable them in the PR, but for those tests where the only issue is leaving behind temporary files, can we try to enable them in this PR?  If the temporary files somehow cause a problem, we'll see it in the CI run for this PR.

File: python/ray/_private/runtime_env/pip.py
Comment (jbedorf) [reply python/ray/_private/runtime_env/pip.py:507]: Yeah I was thinking more about it. A cleaner solution would to have always add lists to the `command_prefix` item and then do the combination of the items to string for Linux/Mac and keep it as list items for Windows. So we would have to update it for each of the three plugins.

File: python/ray/_private/test_utils.py
Comment (jbedorf) [reply python/ray/_private/test_utils.py:1251]: Yep, will update.

**Comment by jbedorf** [comment]
> 
> > As such not all tests are enabled for Windows as they would keep failing, and leave behind sizeable temporary files.
> 
> Certainly if they're failing we shouldn't enable them in the PR, but for those tests where the only issue is leaving behind temporary files, can we try to enable them in this PR? If the temporary files somehow cause a problem, we'll see it in the CI run for this PR.

The tests will pass when adding the same construct as added in some of the other tests:

```
        whitelist = get_local_file_whitelist(cluster, option)
        wait_for_condition(lambda: check_local_files_gced(cluster, whitelist))
```

Basically ignore the top level folder in the `conda` subfolder. It worked fine, but because the temporary session folders are not cleaned during the tests there will be increased disk usage due to various DLLs and exe files that are left behind in the conda folder. For the `workdir` folders that was not an issue as those were empty. I can enable the tests and hope the CI servers don't run out of disk space next time the tests run 😬

**Comment by jbedorf** [comment]
@architkulkarni Has something changed with the buildkite settings? After updating the code and importing the latest master I'm unable to view the build results/details. This makes it impossible to see why some of the steps failed. 

Looking at the links it appears the address previously was `ray-project/ray-builders-pr/..` and now `ray-project/oss-ci-build-pr`

**Comment by architkulkarni** [comment]
@jbedorf Not 100% sure if it's related, but you're right that there was a recent change to our CI pipeline, but it shouldn't require any special action on our part.  I would suggest merging the latest master but it looks like you've already done that.

Which parts can you no longer see? From what I can tell at https://buildkite.com/ray-project/oss-ci-build-pr/builds/396
Screen Shot 2022-09-26 at 8 27 56 AM some runtime env tests are failing on linux/mac, which should be reproducible locally.

**Comment by jbedorf** [comment]
I see, I guess the project settings have changed regarding anonymous access. When I click on the current links it tells me I need a buildkite account. Whereas previous builds could be accessed by anyone. You can see it by opening the buildkite links in an incognito window. 

Anyway, I'll update my Linux environment and take a look at the failing tests.

**Comment by pcmoritz** [comment]
@jbedorf Thanks for your feedback about the buildkite visibility, this is not intended. We are looking into fixing it!

cc @simon-mo @thomasdesr

**Comment by pcmoritz** [comment]
I believe this is fixed now -- I tried it out via incognito mode :)

Thanks @simon-mo and @krfricke for fixing!

**Comment by jbedorf** [comment]
> I believe this is fixed now -- I tried it out via incognito mode :)
> 
> Thanks @simon-mo and @krfricke for fixing!

Yep, it works now. Thanks!

**Comment by jbedorf** [comment]
@architkulkarni 
Please have a another look. The requested changes are made and some more tests are enabled. All related tests do pass, there are some failing tests that appear to either be known issues and/or failing for other PRs as well.

For example:
* https://github.com/ray-project/ray/pull/28825
* `python/ray/tests/test_client_proxy.py::test_delay_in_rewriting_environment `  (Which passes locally)

**Comment by architkulkarni** [comment]
https://flaky-tests.ray.io/
Windows test_traceback broken on master 
Wheels and Jars timeout likely unrelated
Linkcheck failure unrelated (no doc changes)
test_usage_stats broken on master
test_client_proxy broken recently on master

**Review by architkulkarni — APPROVED** [review]
Looks good to me!  Before I merge this, it would be good to have a review from Windows expert @mattip who should be back from OOO soon.

File: python/ray/_private/runtime_env/context.py
Comment (mattip) [inline python/ray/_private/runtime_env/context.py:86]: There are some slight differences when running with `shell=True`. Is this intentional?

File: python/ray/tests/test_runtime_env_complicated.py
Comment (mattip) [inline python/ray/tests/test_runtime_env_complicated.py:982]: +1, this is a welcome change

**Review by mattip — APPROVED** [review]
This is nice. There is only once small nit with changing `shell=True` but I think that is OK. I like that there are more tests enabled and the use of pytest for the tmpdir fixture.

File: python/ray/_private/runtime_env/context.py
Comment (jbedorf) [reply python/ray/_private/runtime_env/context.py:86]: I found that the tests ran more stable with `shell=True`. I honestly don't understand exactly why as the tests that fail with `shell=False` often pass when I run them individually. Maybe it has to do with the way pytest runs/keeps the state between runs, but I couldn't pin it down exactly.

**Pull request merged**
